src/main/java/com/google/sps/servlets/StudyServlet.java

```diff
@@ -0,0 +1,169 @@
+package com.google.sps.servlets;
+
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import com.google.gson.Gson;
+
+import com.google.sps.data.Card;
+import com.google.sps.tool.WordSearch;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.Collections;
+
+/**
+ * Handles the study mode questions and responses by adjusting and
+ * sorting a card's based on their familarity score and last time seen.
+ */
+@WebServlet("/study")
+public class StudyServlet extends HttpServlet {
+  public class QuizCard {
+    private String quizWord;
+    private String cardKey;
+    private ArrayList<String> possibleResponses = new ArrayList<>();
+    private String correctAnswer;
+
+    public QuizCard(Card card) {
+      this.quizWord = card.getTextTranslated();
+      this.cardKey = card.getCardKey();
+      this.correctAnswer = card.getRawText().toLowerCase();
+      this.possibleResponses = WordSearch.generateWordOptions(this.correctAnswer);
+    }
+  }
+
+  class SortByFamilarity implements Comparator<Card> {
+    public int compare(Card a, Card b) {
+      return (int) (a.getFamilarityScore() - b.getFamilarityScore());
+    }
+  }
+
+  private int numOfCardsPerRound = 5;
+  private int maxNumOfRounds = 4;
+
+  // Returns QuizCard's that are sorted by familarity score as json
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    List<Card> userCards = new ArrayList<>();
+
+    if (userService.isUserLoggedIn()) {
+      String folderKey = request.getParameter("folderKey");
+      Query cardQuery = new Query("Card").setAncestor(KeyFactory.stringToKey(folderKey));
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      PreparedQuery results = datastore.prepare(cardQuery);
+
+      if (results != null) {
+        for (Entity entity : results.asIterable()) {
+          // If the card has no familarity score then it's set to the default value 
+          userCards.add(initializeCard(entity));
+        }
+      }
+      // Sorts cards in increasing order of familarity score
+      Collections.sort(userCards, new SortByFamilarity());
```

Comment:
you can replace this line with the lambda code from the above comment

Reply:
done.